### ISSUE TITLE
Expliquer chaque fonction autorisé

### ISSUE THREAD

**Issue by sylvainnicolet** [lorobert42/webserv#1]
# **accept**

accept a new connection on a socket.

```cpp
#include <sys/socket.h>

int accept(int socket, struct sockaddr *restrict address,
       socklen_t *restrict address_len);
```

**DESCRIPTION**

> The accept() function shall extract the first connection on the queue of pending connections, create a new socket with the same socket type protocol and address family as the specified socket, and allocate a new file descriptor for that socket. The file descriptor shall be allocated as described in File Descriptor Allocation.
> 
> 
> The *accept*() function takes the following arguments:
> 
> *socket*Specifies a socket that was created with *[[socket*()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/socket.html)](https://pubs.opengroup.org/onlinepubs/9699919799/functions/socket.html), has been bound to an address with *[[bind*()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/bind.html)](https://pubs.opengroup.org/onlinepubs/9699919799/functions/bind.html), and has issued a successful call to *[[listen*()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/listen.html)](https://pubs.opengroup.org/onlinepubs/9699919799/functions/listen.html).*address*Either a null pointer, or a pointer to a **sockaddr** structure where the address of the connecting socket shall be returned.*address_len*Either a null pointer, if *address* is a null pointer, or a pointer to a **socklen_t** object which on input specifies the length of the supplied **sockaddr** structure, and on output specifies the length of the stored address.
> 
> If *address* is not a null pointer, the address of the peer for the accepted connection shall be stored in the **sockaddr** structure pointed to by *address*, and the length of this address shall be stored in the object pointed to by *address_len*.
> 
> If the actual length of the address is greater than the length of the supplied **sockaddr** structure, the stored address shall be truncated.
> 
> If the protocol permits connections by unbound clients, and the peer is not bound, then the value stored in the object pointed to by *address* is unspecified.
> 
> If the listen queue is empty of connection requests and O_NONBLOCK is not set on the file descriptor for the socket, *accept*() shall block until a connection is present. If the *[listen*()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/listen.html) queue is empty of connection requests and O_NONBLOCK is set on the file descriptor for the socket, *accept*() shall fail and set *errno* to [EAGAIN] or [EWOULDBLOCK].
> 
> The accepted socket cannot itself accept more connections. The original socket remains open and can accept more connections.
> 

**RETURN VALUE**

> Upon successful completion, accept() shall return the non-negative file descriptor of the accepted socket. Otherwise, -1 shall be returned, errno shall be set to indicate the error, and any object pointed to by address_len shall remain unchanged.
> 

# **access**

access, faccessat - determine accessibility of a file descriptor

```cpp
#include <unistd.h>

int access(const char *path, int amode);
```

**DESCRIPTION**

> The access() function shall check the file named by the pathname pointed to by the path argument for accessibility according to the bit pattern contained in amode. The checks for accessibility (including directory permissions checked during pathname resolution) shall be performed using the real user ID in place of the effective user ID and the real group ID in place of the effective group ID.
> 
> 
> The value of *amode* is either the bitwise-inclusive OR of the access permissions to be checked (R_OK, W_OK, X_OK) or the existence test (F_OK).
> 
> If any access permissions are checked, each shall be checked individually, as described in XBD *[File Access Permissions](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_05)*, except that where that description refers to execute permission for a process with appropriate privileges, an implementation may indicate success for X_OK even if execute permission is not granted to any user.
> 
> The *faccessat*() function, when called with a *flag* value of zero, shall be equivalent to the *access*() function, except in the case where *path* specifies a relative path. In this case the file whose accessibility is to be determined shall be located relative to the directory associated with the file descriptor *fd* instead of the current working directory. If the access mode of the open file description associated with the file descriptor is not O_SEARCH, the function shall check whether directory searches are permitted using the current permissions of the directory underlying the file descriptor. If the access mode is O_SEARCH, the function shall not perform the check.
> 
> If *faccessat*() is passed the special value AT_FDCWD in the *fd* parameter, the current working directory shall be used and, if flag is zero, the behavior shall be identical to a call to *access*().
> 
> Values for *flag* are constructed by a bitwise-inclusive OR of flags from the following list, defined in *[<fcntl.h>](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/fcntl.h.html)*:
> 
> AT_EACCESSThe checks for accessibility (including directory permissions checked during pathname resolution) shall be performed using the effective user ID and group ID instead of the real user ID and group ID as required in a call to *access*().
> 

**RETURN VALUE**

> Upon successful completion, these functions shall return 0. Otherwise, these functions shall return -1 and set errno to indicate the error.

# **bind**

bind -- bind a name to a socket

```cpp
#include <sys/socket.h>

int
bind(int socket, const struct sockaddr *address, socklen_t address_len);
```

**DESCRIPTION**
> bind() assigns a name to an unnamed socket.  When a socket is created with socket(2) it exists in a name space (address family) but has no name assigned.
> bind() requests that address be assigned to the socket.

**NOTES**
> Binding a name in the UNIX domain creates a socket in the file system that must be deleted by the caller when it is no longer needed (using unlink(2)).
> The rules used in name binding vary between communication domains.  Consult the manual entries in section 4 for detailed information.

**RETURN VALUES**
> Upon successful completion, a value of 0 is returned.  Otherwise, a value of -1 is returned and the global integer variable errno is set to indicate the error.

# **close**

close - close a file descriptor

```cpp
#include <unistd.h>

int close(int fildes);
```

**DESCRIPTION**

> The close() function shall deallocate the file descriptor indicated by fildes. To deallocate means to make the file descriptor available for return by subsequent calls to open() or other functions that allocate file descriptors. All outstanding record locks owned by the process on the file associated with the file descriptor shall be removed (that is, unlocked).
If close() is interrupted by a signal that is to be caught, it shall return -1 with errno set to [EINTR] and the state of fildes is unspecified. If an I/O error occurred while reading from or writing to the file system during close(), it may return -1 with errno set to [EIO]; if this error is returned, the state of fildes is unspecified.
When all file descriptors associated with a pipe or FIFO special file are closed, any data remaining in the pipe or FIFO shall be discarded.
When all file descriptors associated with an open file description have been closed, the open file description shall be freed.
If the link count of the file is 0, when all file descriptors associated with the file are closed, the space occupied by the file shall be freed and the file shall no longer be accessible.
If a STREAMS-based fildes is closed and the calling process was previously registered to receive a SIGPOLL signal for events associated with that STREAM, the calling process shall be unregistered for events associated with the STREAM. The last close() for a STREAM shall cause the STREAM associated with fildes to be dismantled. If O_NONBLOCK is not set and there have been no signals posted for the STREAM, and if there is data on the module's write queue, close() shall wait for an unspecified time (for each module and driver) for any output to drain before dismantling the STREAM. The time delay can be changed via an I_SETCLTIME ioctl() request. If the O_NONBLOCK flag is set, or if there are any pending signals, close() shall not wait for output to drain, and shall dismantle the STREAM immediately.
If the implementation supports STREAMS-based pipes, and fildes is associated with one end of a pipe, the last close() shall cause a hangup to occur on the other end of the pipe. In addition, if the other end of the pipe has been named by fattach(), then the last close() shall force the named end to be detached by fdetach(). If the named end has no open file descriptors associated with it and gets detached, the STREAM associated with that end shall also be dismantled.
If fildes refers to the master side of a pseudo-terminal, and this is the last close, a SIGHUP signal shall be sent to the controlling process, if any, for which the slave side of the pseudo-terminal is the controlling terminal. It is unspecified whether closing the master side of the pseudo-terminal flushes all queued input and output.
If fildes refers to the slave side of a STREAMS-based pseudo-terminal, a zero-length message may be sent to the master.
When there is an outstanding cancelable asynchronous I/O operation against fildes when close() is called, that I/O operation may be canceled. An I/O operation that is not canceled completes as if the close() operation had not yet occurred. All operations that are not canceled shall complete as if the close() blocked until the operations completed. The close() operation itself need not block awaiting such I/O completion. Whether any I/O operation is canceled, and which I/O operation may be canceled upon close(), is implementation-defined
If a shared memory object or a memory mapped file remains referenced at the last close (that is, a process has it mapped), then the entire contents of the memory object shall persist until the memory object becomes unreferenced. If this is the last close of a shared memory object or a memory mapped file and the close results in the memory object becoming unreferenced, and the memory object has been unlinked, then the memory object shall be removed.
If fildes refers to a socket, close() shall cause the socket to be destroyed. If the socket is in connection-mode, and the SO_LINGER option is set for the socket with non-zero linger time, and the socket has untransmitted data, then close() shall block for up to the current linger interval until all data is transmitted.
> 

**RETURN VALUE**

> Upon successful completion, 0 shall be returned; otherwise, -1 shall be returned and errno set to indicate the error.
>

# **closedir**
opendir, fdopendir, readdir, readdir_r, telldir, seekdir, rewinddir, closedir, dirfd -- directory operations

```cpp
#include <dirent.h>

int
closedir(DIR *dirp);
```

**DESCRIPTION**
> The closedir() function closes the named directory stream and frees the structure associated with the dirp pointer, returning 0 on success.  On failure, -1 is returned and the global variable errno is set to indicate the error.

# **connect**
connect -- initiate a connection on a socket

```cpp
#include <sys/types.h>
#include <sys/socket.h>

int
connect(int socket, const struct sockaddr *address, socklen_t address_len);
```

**DESCRIPTION**
> The parameter socket is a socket.  If it is of type SOCK_DGRAM, this call specifies the peer with which the socket is to be associated; this address is that to which datagrams are to be sent, and the only address from which datagrams are to be received.  If the socket is of type SOCK_STREAM, this call attempts to make a connection to another socket.  The other socket is specified by address, which is an address in the communications space of the socket.
>
> Each communications space interprets the address parameter in its own way.  Generally, stream sockets may successfully connect() only once; datagram sockets may use connect() multiple times to change their association.  Datagram sockets may dissolve the association by calling disconnectx(2), or by connecting to an invalid address, such as a null address or an address with the address family set to AF_UNSPEC (the error EAFNOSUPPORT will be harmlessly returned).

**RETURN VALUES**
> Upon successful completion, a value of 0 is returned.  Otherwise, a value of -1 is returned and the global integer variable errno is set to indicate the error.

# **dup, dup2**
dup, dup2 -- duplicate an existing file descriptor

```cpp
#include <unistd.h>

int
dup(int fildes);

int
dup2(int fildes, int fildes2);
```

**DESCRIPTION**
> dup() duplicates an existing object descriptor and returns its value to the calling process (fildes2 = dup(fildes)).  The argument fildes is a small non-negative integer index in the per-process descriptor table.  The value must be less than the size of the table, which is returned by getdtablesize(2).  The new descriptor returned by the call is the lowest numbered descriptor currently not in use by the process.
>
> The object referenced by the descriptor does not distinguish between fildes and fildes2 in any way.  Thus if fildes2 and fildes are duplicate references to an open file, read(2), write(2) and lseek(2) calls all move a single pointer into the file, and append mode, non-blocking I/O and asynchronous I/O options are shared between the references.  If a separate pointer into the file is desired, a different object reference to the file must be obtained by issuing an additional open(2) call.  The close-on-exec flag on the new file descriptor is unset.
>
> In dup2(), the value of the new descriptor fildes2 is specified.  If fildes and fildes2 are equal, then dup2() just returns fildes2; no other changes are made to the existing descriptor.  Otherwise, if descriptor fildes2 is already in use, it is first deallocated as if a close(2) call had been done first.

**RETURN VALUES**
> Upon successful completion, the new file descriptor is returned.  Otherwise, a value of -1 is returned and the global integer variable errno is set to indicate the error.

# **epoll** (only [Linux](https://man7.org/linux/man-pages/man7/epoll.7.html))
epoll - I/O event notification facility

```cpp
#include <sys/epoll.h>
```

**DESCRIPTION**
> The epoll API performs a similar task to [poll(2)](https://man7.org/linux/man-pages/man2/poll.2.html): monitoring multiple file descriptors to see if I/O is possible on any of them.  The epoll API can be used either as an edge-triggered or a level-triggered interface and scales well to large numbers of watched file descriptors.
>
> The central concept of the epoll API is the epoll instance, an in-kernel data structure which, from a user-space perspective, can be considered as a container for two lists:
>
> - The interest list (sometimes also called the epoll set): the set of file descriptors that the process has registered an
          interest in monitoring.
>
> - The ready list: the set of file descriptors that are "ready" for I/O.  The ready list is a subset of (or, more precisely, a set of references to) the file descriptors in the interest list.  The ready list is dynamically populated by the kernel as a result of I/O activity on those file descriptors.
>
> The following system calls are provided to create and manage an epoll instance:
>
> - [epoll_create(2)](https://man7.org/linux/man-pages/man2/epoll_create.2.html) creates a new epoll instance and returns a file descriptor referring to that instance.  (The more recent [epoll_create1(2)](https://man7.org/linux/man-pages/man2/epoll_create1.2.html) extends the functionality of [epoll_create(2)](https://man7.org/linux/man-pages/man2/epoll_create.2.html).)
>
> - Interest in particular file descriptors is then registered via [epoll_ctl(2)](https://man7.org/linux/man-pages/man2/epoll_ctl.2.html), which adds items to the interest list of the epoll instance.
>
> - [epoll_wait(2)](https://man7.org/linux/man-pages/man2/epoll_wait.2.html) waits for I/O events, blocking the calling thread if no events are currently available.  (This system call can be thought of as fetching items from the ready list of the epoll instance.)

# **epoll_create**
epoll_create, epoll_create1 - open an epoll file descriptor

```cpp
#include <sys/epoll.h>

int epoll_create(int size);
```

**DESCRIPTION**
> epoll_create() creates a new [epoll(7)](https://man7.org/linux/man-pages/man7/epoll.7.html) instance.  Since Linux 2.6.8, the size argument is ignored, but must be greater than zero; see HISTORY.
>
> epoll_create() returns a file descriptor referring to the new epoll instance.  This file descriptor is used for all the subsequent calls to the epoll interface.  When no longer required, the file descriptor returned by epoll_create() should be closed by using [close(2)](https://man7.org/linux/man-pages/man2/close.2.html).  When all file descriptors referring to an epoll instance have been closed, the kernel destroys the instance and releases the associated resources for reuse.

**RETURN VALUE**
> On success, these system calls return a file descriptor (a nonnegative integer).  On error, -1 is returned, and [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to indicate the error.

# **epoll_ctl**
epoll_ctl - control interface for an epoll file descriptor

```cpp
#include <sys/epoll.h>

int epoll_ctl(int epfd, int op, int fd, struct epoll_event *_Nullable event);
```

**DESCRIPTION**
> This system call is used to add, modify, or remove entries in the interest list of the [epoll(7)](https://man7.org/linux/man-pages/man7/epoll.7.html) instance referred to by the file descriptor epfd.  It requests that the operation op be performed for the target file descriptor, fd.
>
> Valid values for the op argument are (see [here](https://man7.org/linux/man-pages/man2/epoll_ctl.2.html))

**RETURN VALUE**
> When successful, epoll_ctl() returns zero.  When an error occurs, epoll_ctl() returns -1 and [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to indicate the error.

# **epoll_wait**
epoll_wait, epoll_pwait, epoll_pwait2 - wait for an I/O event on an epoll file descriptor

```cpp
#include <sys/epoll.h>

int epoll_wait(int epfd, struct epoll_event *events, int maxevents, int timeout);
```

**DESCRIPTION**
> The epoll_wait() system call waits for events on the [epoll(7)](https://man7.org/linux/man-pages/man7/epoll.7.html) instance referred to by the file descriptor epfd.  The buffer pointed to by events is used to return information from the ready list about file descriptors in the interest list that have some events available.  Up to maxevents are returned by epoll_wait(). The maxevents argument must be greater than zero.
>
> The timeout argument specifies the number of milliseconds that epoll_wait() will block.  Time is measured against the CLOCK_MONOTONIC clock.
>
> A call to epoll_wait() will block until either:
>
> -  a file descriptor delivers an event;
>
> -  the call is interrupted by a signal handler; or
>
> -  the timeout expires.
>
> Note that the timeout interval will be rounded up to the system clock granularity, and kernel scheduling delays mean that the blocking interval may overrun by a small amount.  Specifying a timeout of -1 causes epoll_wait() to block indefinitely, while specifying a timeout equal to zero cause epoll_wait() to return immediately, even if no events are available.
>
> The struct epoll_event is described in [epoll_event(3type)](https://man7.org/linux/man-pages/man3/epoll_event.3type.html).
>
> The data field of each returned epoll_event structure contains the same data as was specified in the most recent call to [epoll_ctl(2)](https://man7.org/linux/man-pages/man2/epoll_ctl.2.html) (EPOLL_CTL_ADD, EPOLL_CTL_MOD) for the corresponding open file descriptor.
>
> The events field is a bit mask that indicates the events that have occurred for the corresponding open file description.  See [epoll_ctl(2)](https://man7.org/linux/man-pages/man2/epoll_ctl.2.html) for a list of the bits that may appear in this mask.

**RETURN VALUE**
> On success, epoll_wait() returns the number of file descriptors ready for the requested I/O, or zero if no file descriptor became ready during the requested timeout milliseconds.  On failure, epoll_wait() returns -1 and [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to indicate the error.

# **errno**
errno - number of last error

```cpp
#include <errno.h>
```

**DESCRIPTION**
> The  <errno.h>  header  file defines the integer variable errno, which is set by system calls and some library functions in the event of an error to indicate what went wrong.
>
> The value in errno is significant only when the return value of the call indicated an  error  (i.e.,  -1  from most  system calls; -1 or NULL from most library functions); a function that succeeds is allowed to change errno.  The value of errno is never set to zero by any system call or library function.
>
> For some system calls and library functions (e.g., getpriority(2)), -1 is a valid return on success. In such cases, a successful return can be distinguished from an error return by setting errno to zero before the call, and then, if the call returns a status that indicates that an error may have occurred, checking to see if  errno has a nonzero value.

# **execve**
execve - execute program

```cpp
#include <[unistd.h](https://linux.die.net/include/unistd.h)>

int execve(const char *filename, char *const argv[], char *const envp[]);
```

**DESCRIPTION**
> execve() executes the program pointed to by filename. filename must be either a binary executable, or a script starting with a line of the form:
>
>`#! interpreter [optional-arg]`
>
> For details of the latter case, see "Interpreter scripts" below.
>
> argv is an array of argument strings passed to the new program. By convention, the first of these strings should contain the filename associated with the file being executed. envp is an array of strings, conventionally of the form key=value, which are passed as environment to the new program. Both argv and envp must be terminated by a NULL pointer. The argument vector and environment can be accessed by the called program's main function, when it is defined as:
>
> `int main(int argc, char *argv[], char *envp[])`
>
> execve() does not return on success, and the text, data, bss, and stack of the calling process are overwritten by that of the program loaded. 

# **fcntl**
fcntl - manipulate file descriptor

```cpp
#include <[unistd.h](https://linux.die.net/include/unistd.h)>
#include <[fcntl.h](https://linux.die.net/include/fcntl.h)>

int fcntl(int fd, int cmd, ... /* arg */ );
```

**DESCRIPTION**
> fcntl() performs one of the operations described [here](https://linux.die.net/man/2/fcntl) on the open file descriptor fd. The operation is determined by cmd.
>
> fcntl() can take an optional third argument. Whether or not this argument is required is determined by cmd. The required argument type is indicated in parentheses after each cmd name (in most cases, the required type is int, and we identify the argument using the name arg), or void is specified if the argument is not required. 

**fork** :

# **freeaddrinfo, gai_strerror, getaddrinfo**
getaddrinfo, freeaddrinfo, gai_strerror - network address and service translation

```cpp
#include <[sys/types.h](https://linux.die.net/include/sys/types.h)>
#include <[sys/socket.h](https://linux.die.net/include/sys/socket.h)>
#include <[netdb.h](https://linux.die.net/include/netdb.h)>

int getaddrinfo(const char *node, const char *service,
                const struct addrinfo *hints,
                struct addrinfo **res);

void freeaddrinfo(struct addrinfo *res);

const char *gai_strerror(int errcode);
```

**DESCRIPTION**
> Given node and service, which identify an Internet host and a service, getaddrinfo() returns one or more addrinfo structures, each of which contains an Internet address that can be specified in a call to [bind](https://linux.die.net/man/2/bind)(2) or [connect](https://linux.die.net/man/2/connect)(2). The getaddrinfo() function combines the functionality provided by the [gethostbyname](https://linux.die.net/man/3/gethostbyname)(3) and [getservbyname](https://linux.die.net/man/3/getservbyname)(3) functions into a single interface, but unlike the latter functions, getaddrinfo() is reentrant and allows programs to eliminate IPv4-versus-IPv6 dependencies.
>
> The addrinfo structure used by getaddrinfo() contains the following fields:
>
```cpp
    struct addrinfo {
        int              ai_flags;
        int              ai_family;
        int              ai_socktype;
        int              ai_protocol;
        socklen_t        ai_addrlen;
        struct sockaddr *ai_addr;
        char            *ai_canonname;
        struct addrinfo *ai_next;
    };
```
>
> The hints argument points to an addrinfo structure that specifies criteria for selecting the socket address structures returned in the list pointed to by res. If hints is not NULL it points to an addrinfo structure whose ai_family, ai_socktype, and ai_protocol specify criteria that limit the set of socket addresses returned by getaddrinfo(), as explained [here](https://linux.die.net/man/3/freeaddrinfo)

**getprotobyname** :

**getsockname** :

**htonl** :

**htons** :

**kevent** :

**kill** :

**kqueue** :

**kqueue** :

**listen** :

**ntohl** :

**ntohs** :

**opendir** :

**pipe** :

**poll** :

**read** :

**readdir** :

**recv** :

**select** :

**send** :

**setsockopt** :

**signal** :

**socket** :

**stat** :

**strerror** :

**waitpid** :

**write** :
